source/globalCommands.py

```diff
@@ -4192,44 +4192,61 @@ def script_reportLinkDestination(
 		except RuntimeError:
 			log.debugWarning("Unable to get the caret position.", exc_info=True)
 			ti: textInfos.TextInfo = api.getFocusObject().makeTextInfo(textInfos.POSITION_FIRST)
-		ti.expand(textInfos.UNIT_CHARACTER)
-		obj: NVDAObject = ti.NVDAObjectAtStart
+		try:
+			link = ti._getLinkDataAtCaretPosition()
+		except NotImplementedError:
+			link = self._getLinkDataAtCaretPosition(ti)
 		presses = scriptHandler.getLastScriptRepeatCount()
-		if obj.role == controlTypes.role.Role.GRAPHIC and (
-			obj.parent and obj.parent.role == controlTypes.role.Role.LINK
-		):
-			# In Firefox, graphics with a parent link also expose the parents link href value.
-			# In Chromium, the link href value must be fetched from the parent object. (#14779)
-			obj = obj.parent
-		if (
-			obj.role == controlTypes.role.Role.LINK  # If it's a link, or
-			or controlTypes.state.State.LINKED in obj.states  # if it isn't a link but contains one
-		):
-			linkDestination = obj.value
-			if linkDestination is None:
+		if link:
+			if link.destination is None:
 				# Translators: Informs the user that the link has no destination
 				ui.message(_("Link has no apparent destination"))
 				return
 			if (
 				presses == 1  # If pressed twice, or
 				or forceBrowseable  # if a browseable message is preferred unconditionally
 			):
+				text = link.displayText
+				if text is None:
+					# Translators: Title of the browseable message when requesting the destination of a graphical link.
+					text = _("Graphic")
 				ui.browseableMessage(
-					linkDestination,
+					link.destination,
 					# Translators: Informs the user that the window contains the destination of the
 					# link with given title
-					title=_("Destination of: {name}").format(name=obj.name),
-					closeButton=True,
-					copyButton=True,
+					title=_("Destination of: {name}").format(
+						name=text,
+						closeButton=True,
+						copyButton=True,
+					),
 				)
 			elif presses == 0:  # One press
-				ui.message(linkDestination)  # Speak the link
+				ui.message(link.destination)  # Speak the link
 			else:  # Some other number of presses
 				return  # Do nothing
 		else:
 			# Translators: Tell user that the command has been run on something that is not a link
 			ui.message(_("Not a link."))
 
+	def _getLinkDataAtCaretPosition(self, ti: textInfos.TextInfo) -> textInfos._Link | None:
```

Comment:
>Let me know if you wish me to move the original code in suitable text info object(s) rather than to keep it in the global commands class. Moving this code would surely be cleaner regarding the code organization; but I am more likely to break existing use cases. If you push for moving this code, have you an idea where I should move it? (advice appreciated)

I don't have strong feelings on this, but I think `utils.urlUtils` is probably the best place. Would this break anything?

Reply:
Sorry, I have not been clear enough. My question was the following:

For now, in `script_reportLinkDestination`:
* For Word/Outlook, the link is retrieved through their document's text infos, i.e. `link = ti._getLinkDataAtCaretPosition()`
* For UIA or Chrome/Firefox, the link is retrieved with a generic function i.e. `GlobalCommands._getLinkDataAtCaretPosition` ("`except NotImplemented`" clause in the script)

If we move the code of `_getLinkDataAtCaretPosition` from `GlobalCommands` to `utils.urlUtils` this remains a global function; it is not better for code structure.

My suggestion was to move this code from a global function to a text infos implementation that would be valid for UIA, Chrome and Firefox.

If you have no strong recommendation on this, I'd better keep this piece of code here as a helper function of the script, next to it...

Reply:
>If we move the code of _getLinkDataAtCaretPosition from GlobalCommands to utils.urlUtils this remains a global function; it is not better for code structure.

I think its better for code structure to take helper functions out of the already large GlobalCommands. However I agree moving it to textinfos is even better. Is there anything stopping you from doing that?

Reply:
I have moved this code to `TextInfo` in commit a5ab45913a1fd535d03ce32f20b496f13c718f23.

This PR is ready again for review.